Testbench: Add different in and out channels count and improve shell helper scripts for components

singalsu opened the pull request.

singalsu commented:
@marc-hb What do you think of the component run scripts? I will do later a shellcheck fix round for details but does the overall approach look sane to you.

I did this work to help make more advanced automatic tests for beamformer and sound direction tracking. The test script (not yet this patch) captures debug trace output from testbench and parses and checks some internal state variables that way during the test run.

marc-hb commented on scripts/host-testbench.sh:
```diff
@@ -56,64 +105,61 @@ rm -rf ./*.raw
+# test with tdfb
+ALGO=tdfb
+CMD="./${ALGO}_run.sh -v -b 16 zeros_in.raw ${ALGO}_out.raw"
+test_comp "$CMD" "$ALGO"
```
This should be a loop.

singalsu replied:
This part of patch was removed, I'll do another PR to add more components to this test.

marc-hb commented on scripts/host-testbench.sh:
```diff
@@ -44,6 +44,55 @@ function srcsize() {
   echo "$OUTPUT_SIZE"
 }
 
+function test_comp() {
```
The `function` keyword is not required and less portable.

singalsu replied:
ditto

marc-hb reviewed (commented):
Too much code and layers of indirections. Many of the top-level scripts are just configuration files in disguise. The double level of `getopt` is especially hurting.

My recommendation: add a new `./comp_run.sh --test_config` option and get rid of most top level scripts. Keep it simple: have `./comp_run.sh` merely `source` the config file which defines a bunch of test parameters as shell variables.

marc-hb commented on tools/test/audio/asrc_run.sh:
```diff
@@ -2,25 +2,65 @@
-    DIRECTION=playback
-
-    ./comp_run.sh $COMP $DIRECTION "$@"
+    ./comp_run.sh $VALGRIND $BITS $CHANNELS $RATE_IN $RATE_OUT $COMP $DIRECTION "$@"
```
You don't need to parse arguments in the first script where they are not used to merely pass them to the next script where they will be parsed again. That's a ton of duplication. As long as some order is enforced, the first script can instead use `shift` to remove its own arguments and then just pass-through the rest using `"$@"`

"$@" examples:
https://github.com/thesofproject/sof/blob/b95d995/scripts/docker-run.sh
https://github.com/thesofproject/sof/blob/b95d995bd2fb1f4f/scripts/xtensa-build-zephyr.sh#L385

and more in git grep -C 5 -F '"$@"'

But again: switching to config files and a single layer is much better.

singalsu replied:
Thanks for tips, yep now using config files.

lyakh commented on scripts/host-testbench.sh:
```diff
@@ -44,6 +44,55 @@ function srcsize() {
+    OUTFN=${COMP}_out.raw
+    echo "=========================================================="
+    echo "test $COMP with ${CMD}"
+    if ${CMD} &> $LOGFN; then
```
indentation looks weird at least for me here in firefox - could you check please? First indentation level lines in the function seem to be indented with spaces, second level (within `if` etc.) - with TABs. Is this intended?

singalsu replied:
I removed this part of patch

lyakh commented on scripts/host-testbench.sh:
```diff
@@ -44,6 +44,55 @@ function srcsize() {
+	exit 1
+    fi
+
+    if comparesize "$(srcsize ${INPUT_FILE_SIZE} $3 $4)" "$(filesize $OUTFN)";then
```
do I see it correctly, that the only difference between these two functions is the first parameter of `comparesize`? Maybe they could be further unified then to be called `x(cmd, comp, ${INPUT_FILE_SIZE})` and `x(cmd, comp, $(srcsize ${INPUT_FILE_SIZE} $3 $4))`?

singalsu replied:
Maybe I'll do another PR for test case increase for new components, @marc-hb proposed that this PR would be min changes only to support the additional testbench command line parameters. So this script would not change at all.

singalsu replied:
Thanks, I will use this idea later when I add more components to default test in CI

marc-hb commented:
> @marc-hb proposed that this PR would be min changes only to support the additional testbench command line parameters. So this script would not change at all.

IMHO it's more effective to submit small, less controversial PRs very quick and easy to review - while still having a clear long term direction in mind. This also provides much more test coverage, easier bisects, etc. I think it's called "Continuous Integration" ;-)

singalsu commented on tools/test/audio/test_utils/test_run.m:
```diff
@@ -30,24 +30,64 @@
+fprintf(fh, 'FN_IN=\"%s\"\n', test.fn_in);
+fprintf(fh, 'FN_OUT=\"%s\"\n', test.fn_out);
+
+if isfield(test, 'run_opts')
```
this should be test for extra_opts

singalsu replied:
This is now fixed.

marc-hb commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,87 @@
+shift $((OPTIND -1))
+
+if [[ -z $SOURCED_CONFIG ]]; then
+    if [ $# -ne 8 ]; then
```
"do || die" / assert style:
```suggestion
    [ -n "$SOURCED_CONFIG" ] || [ $# -eq 8 ] || {
        usage
        exit 1
        }
```

singalsu replied:
This maybe works, I haven't tested with Matlab tests, but I don't think it improves readability:

```
    if ! "$SOURCED_CONFIG"; then
	[ $# -eq 8 ] || {
	    usage "$0"
	    exit 1
	}
	
	COMP=$1
	DIRECTION=$2
	BITS_IN=$3
	BITS_OUT=$4
	FS_IN=$5
	FS_OUT=$6
	FN_IN=$7
	FN_OUT=$8
    fi

```

singalsu replied:
As above now, but I guess it's not what you were suggesting.

marc-hb commented on tools/test/audio/comp_run.sh:
```diff
@@ -35,16 +108,30 @@ else
-valgrind --leak-check=yes --error-exitcode=1 $CMD
+rm -f "$FN_OUT"
+rm -f "$FN_TRACE"
+if [ "$VALGRIND" = "yes" ]; then
```
You don't need the `if VALGRIND` test.

If you drop the `2>` redirection and switch to `--log-file` or something, then you don't need the `if FN_TRACE` either.

You can run this same command with $MAYBE_VALGRIND maybe empty maybe not and it will Just Work:
```
  $MAYBE_VALGRIND $CMD
```
Again `$MAYBE_REDIRECT` does _not_ work, you need the if.

(I'm not suggesting to actually prefix the variable with MAYBE_)

singalsu replied:
LOL -- I might have used maybe :) Good recommendations, thanks!

marc-hb replied:
>  I might have used maybe :) 

Maybe it's better :-)

singalsu replied:
Done with $VALGRIND_CMD $CMD, no --log-file yet.

marc-hb reviewed (changes requested):
Nothing standing out as fundamentally wrong in the shell script code but requesting changes because of the sheer number of recommendations.

singalsu commented:
I've not yet addressed the comments. The previous push added enable of verbose traces for library build for testbench. The extra_opts handling is fixed.

singalsu commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,87 @@
+    echo "FN_IN=input.raw"
+    echo "FN_OUT=output.raw"
+    echo "FN_TRACE=trace.txt"
+    echo "VALGRIND=yes"
```
Done, safe to resolve.

singalsu commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,87 @@
+	    exit 1
+	    ;;
+    esac
+done
```
I did that, hope it's like you had in mind.

singalsu commented on tools/test/audio/comp_run.sh:
```diff
@@ -35,16 +108,30 @@ else
+    if [ -z "$FN_TRACE" ]; then
+	$VALGRIND_CMD $CMD
+    else
+	$VALGRIND_CMD $CMD 2> "$FN_TRACE"
```
Not yet done, valgrind outputs into beginning and end, with experience it's easy to spot which is it. But obviously it's not the case for occasional users. The testbench output is quite a mess today, some trace and markup language prints appear mixed. I don't know who needs the markup. It would need a big cleanup with redirect of everything with simulated time stamps.

singalsu commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,87 @@
+    echo "FS_OUT=48000"
+    echo "FN_IN=input.raw"
+    echo "FN_OUT=output.raw"
+    echo "FN_TRACE=trace.txt"
```
Done, sort of.

singalsu commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,87 @@
+
+shift $((OPTIND -1))
+
+if [[ -z $SOURCED_CONFIG ]]; then
```
Done for SOURCED_CONFIG and VALGRIND.

marc-hb commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,111 @@
+	FS_IN=$5
+	FS_OUT=$6
+	FN_IN=$7
+	FN_OUT=$8
```
shellcheck warnings?

singalsu replied:
Yes, shellcheck thinks I should use `$VALGRIND_CMD "$CMD"`. But in that case valgrind things the whole testbench command is the binary to run. Do you ideas how to do this fix in way that still works?

marc-hb replied:
> Yes, shellcheck thinks I should use $VALGRIND_CMD "$CMD". 

"$CMD" is the tricky one, I wasn't referring to this one. I was only commenting about the easy ones like: `FN_OUT="$8"`.

There are two ways to deal with the `$CMD` quoting problem:
- Add a `# shellcheck disable=SC1234` silencer comment that states "thank you shellcheck but I know what I'm doing"
- Replace $CMD with a bash array `${CMD[@]}` https://mywiki.wooledge.org/BashGuide/Arrays
- Not do anything and not achieve shellcheck perfection. The fewer warnings the better, does not have to be zero.

singalsu replied:
Shellcheck didn't complain about likes of FN_OUT=$8. OK, I'll fix those.

marc-hb reviewed (commented):
Besides some missing quotes (please run shellcheck) the shell script changes look good to me.

I cannot review the matlab and C code unfortunately, not familiar enough with them.

lgirdwood reviewed (approved):
Reviewed the C parts. @ShriramShastry can you review the matlab. Thanks

marc-hb commented on tools/test/audio/comp_run.sh:
```diff
@@ -5,14 +5,114 @@
+{
+    cat <<EOFHELP
+Usage:     $0 <options> <comp direction bits_in bits_out fs_in fs_out input output>
+Example 1: $0 volume playback 16 16 48000 48000 input.raw output.raw
```
Add some `-e` or other option in this example to make it even clearer that combining flags and positional parameters is OK.

marc-hb reviewed (approved):
> Shellcheck didn't complain about likes of FN_OUT=$8. OK, I'll fix those.

Sorry I forgot the first exception in https://mywiki.wooledge.org/Quotes#When_Should_You_Quote.3F
(assignments)

Instead I follow this rule:
https://mywiki.wooledge.org/Quotes#I.27m_Too_Lazy_to_Read.2C_Just_Tell_Me_What_to_Do
> When in doubt, double-quote every expansion in your shell commands.
